src/view/view-3d.cpp

```diff
@@ -223,15 +223,43 @@ wf::pointf_t wf::view_3D::transform_point(
         v.y /= v.w;
     }
 
-    return get_absolute_coords_from_relative(geometry, {v.x, v.y});
+    return get_absolute_coords_from_relative(view->get_wm_geometry(), {v.x, v.y});
 }
 
-/* TODO: is there a way to realiably reverse projective transformations? */
 wf::pointf_t wf::view_3D::untransform_point(wf::geometry_t geometry,
     wf::pointf_t point)
 {
-    return {wf::compositor_core_t::invalid_coordinate,
-        wf::compositor_core_t::invalid_coordinate};
+    auto p  = get_center_relative_coords(view->get_wm_geometry(), point);
+    auto tr = calculate_total_transform();
+
+    /* Since we know that our original z coordinates were zero, we can write a
+     * system of linear equations for the original (x,y) coordinates by writing
+     * out the (x,y,w) components of the transformed coordinate.
```

Comment:
(x,y,z) ?

Reply:
No, in the forward transform, the z coordinate is discarded and w is used for the perspective projection, see [here](https://github.com/WayfireWM/wayfire/blob/d87aba715ff73a45202e3368c0fcde121c220ef0/src/view/view-3d.cpp#L222) -- so we have to use the expression for w here as well.